storage/src/vespa/storage/distributor/pendingmessagetracker.h

```diff
@@ -147,48 +160,167 @@ class PendingMessageTracker : public framework::HtmlStatusReporter {
+
+        // generic emplace, keeping indexes in sync
+        template<typename ...Args>
+        void emplace(Args&&... args)
+        {
+            auto [iter, added] = byMessageIdSet.emplace(std::forward<Args>(args)...);
+            assert(added);
+            const MessageEntry &entry = *iter;
+            byNodeAndBucketSet.insert(&entry);
+            byBucketAndTypeSet.insert(&entry);
+        }
+
+        // remove by key, keeping indexes in sync
+        void remove(const MessageEntry &key) {
+            auto iter = byMessageIdSet.find(key);
+            assert (iter != byMessageIdSet.end());
+            const MessageEntry &entry = *iter;
+            byNodeAndBucketSet.erase(&entry);
+            byBucketAndTypeSet.erase(&entry);
+            byMessageIdSet.erase(iter);
+        }
+
+        // Index wrapping the set using NodeBucketTypeIdComparator
+        struct IndexByNodeAndBucket {
+            using iterator = wrap_set_iterator<ByNodeAndBucketSet::const_iterator>;
+            Messages& _m;
+            auto begin() const { return iterator(_m.byNodeAndBucketSet.begin()); }
+            auto end()   const { return iterator(_m.byNodeAndBucketSet.end()); }
+
+            IndexByNodeAndBucket(Messages& m) : _m(m) {}
+            template<typename Key>
+            std::pair<iterator, iterator> equal_range(Key key) const {
+                auto inner_range = _m.byNodeAndBucketSet.equal_range(key);
+                return std::make_pair(iterator(inner_range.first),
+                                      iterator(inner_range.second));
+            }
+            iterator erase(iterator it) {
+                const MessageEntry& entry = *it;
+                ++it;
+                _m.remove(entry);
+                return it;
+            }
+            void erase(iterator it, iterator to) {
+                while (it != to) {
+                    it = erase(it);
+                }
+            }
+        };
+
+        // Index wrapping the set using BucketTypeNodeIdComparator
+        struct IndexByBucketAndType {
+            using iterator = wrap_set_iterator<ByBucketAndTypeSet::const_iterator>;
+            Messages& _m;
+            auto begin() const { return iterator(_m.byBucketAndTypeSet.begin()); }
+            auto end()   const { return iterator(_m.byBucketAndTypeSet.end()); }
+            std::pair<iterator, iterator> equal_range(BucketKey key) const {
+                auto inner_range = _m.byBucketAndTypeSet.equal_range(key);
+                return std::make_pair(iterator(inner_range.first),
+                                      iterator(inner_range.second));
+            }
+            IndexByBucketAndType(Messages& m) : _m(m) {}
+        };
+
+        // the actual contents:
+        MainSet byMessageIdSet;
+        // sets of pointers:
+        ByNodeAndBucketSet   byNodeAndBucketSet;
+        ByBucketAndTypeSet   byBucketAndTypeSet;
```

Comment:
Consider having these `private` and hide `const&` access behind accessor methods to reduce the chance that anyone accidentally mutates these indexes from the outside, possibly violating core invariants